ergo/distributions/point_density.py

```diff
@@ -0,0 +1,260 @@
+            scale=scale_classes[0].structure((scale_classes, scale_numeric)),
+            normalized=True,
+            traceable=True,
+            cumulative_normed_ps=density_numeric[2],
+        )
+
+    def destructure(self):
+        scale_classes, scale_numeric = self.scale.destructure()
+        class_params = (self.__class__, scale_classes)
+        self_numeric = self.normed_xs, self.normed_densities, self.cumulative_normed_ps
+        numeric_params = (self_numeric, scale_numeric)
+        return (class_params, numeric_params)
+
+    # Create
+
+    @classmethod
+    def from_pairs(
+        cls, pairs, scale: Scale, normalized=False, allow_non_standard_pairs=True
+    ):
+        sorted_pairs = sorted([(v["x"], v["density"]) for v in pairs])
+        xs = np.array([x for (x, density) in sorted_pairs])
+        densities = np.array([density for (x, density) in sorted_pairs])
+        if not normalized:
+            xs = scale.normalize_points(xs)
+            densities = scale.normalize_densities(xs, densities)
+
+        if allow_non_standard_pairs:
+            # interpolate ps at target_xs
+            if not (
+                len(xs) == len(constants.target_xs)
+                and np.isclose(xs, constants.target_xs, rtol=1e-04).all()
+            ):
+                f = interp1d(xs, densities)
+                densities = f(constants.target_xs)
+
+        # Make sure AUC is 1
+        auc = np.sum(densities) / densities.size
+        densities /= auc
+
+        return cls(constants.target_xs, densities, scale=scale, normalized=True)
+
+    @classmethod
+    def from_conditions(
+        cls,
+        *args,
+        fixed_params=None,
+        scale=None,
+        num_points=constants.point_density_default_num_points,
+        **kwargs,
+    ):
+
+        if fixed_params is None:
+            # TODO: Seems weird to denormalize when will be normalized in normalize_fixed_params
+            fixed_params = {"xs": scale.denormalize_points(constants.target_xs)}
+        else:
+            raise Exception("Point Density does not accept custom xs specifications")
+
+        return super(PointDensity, cls).from_conditions(
+            *args, fixed_params=fixed_params, scale=scale, **kwargs
+        )
+
+    @classmethod
+    def from_params(cls, fixed_params, opt_params, scale=Scale(0, 1), traceable=True):
+        xs = fixed_params["xs"]
+        ps = nn.softmax(opt_params) * opt_params.size
+        densities = ps
+        return cls(
+            xs=xs, densities=densities, scale=scale, normalized=True, traceable=True
+        )
+
+    # Optimize Create Helpers
+
+    @staticmethod
+    def initialize_optimizable_params(fixed_params):
+        num_points = fixed_params["xs"].size
+        return onp.full(num_points, 1.0)
+
+    @classmethod
+    def normalize_fixed_params(self, fixed_params, scale):
+        return {"xs": scale.normalize_points(fixed_params["xs"])}
+
+    # Export
+
+    def to_arrays(self, metaculus_denorm=False):
```

Comment:
Is there a more descriptive name for `metaculus_denorm`?

Reply:
Renamed to denorm_xs_only